File: course content/4_mistaken_assumptions/assortative_mating.R
#' Generate allele counts with simulated assortative mating.
#'
#' This function follows a "mixture model" where there's two possibilities. Each locus can be either:
#' - so strongly affected by assortative mating that the two alleles are identical
#' - not affected at all, so the alleles are independent.
#' The chance for the first outcome is the input strength_of_assortative_mating. 
#' 
generateAlleleCounts = function(allele_frequency, strength_of_assortative_mating, n){
  # Check user input -- should be numbers from 0 to 1
  stopifnot(is.numeric(allele_frequency))
  stopifnot(is.numeric(strength_of_assortative_mating))
  stopifnot("strength_of_assortative_mating must be between 0 and 1" = allele_frequency > 0)
  stopifnot("strength_of_assortative_mating must be between 0 and 1" = allele_frequency < 1)
  stopifnot("Allele frequency must be between 0 and 1" = allele_frequency > 0)
  stopifnot("Allele frequency must be between 0 and 1" = allele_frequency < 1)
  extreme_assortative_mating = 2*rbinom(size = 1, prob = allele_frequency, n=n)
  no_assortative_mating = rbinom(size = 2, prob = allele_frequency, n=n)
  ifelse(rbinom(size = 1, prob = strength_of_assortative_mating, n), 
         extreme_assortative_mating,
         no_assortative_mating)
}

#' This function uses the Law of Total Variance to compute the variance of the outcomes from the above mixture model.
#'
#' Law of Total Variance: https://en.wikipedia.org/wiki/Law_of_total_variance
#' 
computeVariance = function(allele_frequency, strength_of_assortative_mating){
  # Check user input -- should be numbers from 0 to 1
  stopifnot(is.numeric(allele_frequency))
  stopifnot(is.numeric(strength_of_assortative_mating))
  stopifnot("strength_of_assortative_mating must be between 0 and 1" = allele_frequency > 0)
  stopifnot("strength_of_assortative_mating must be between 0 and 1" = allele_frequency < 1)
  stopifnot("Allele frequency must be between 0 and 1" = allele_frequency > 0)
  stopifnot("Allele frequency must be between 0 and 1" = allele_frequency < 1)
  variance_if_independent = 2*allele_frequency*(1-allele_frequency)
  variance_if_identical = 4*allele_frequency*(1-allele_frequency)
  strength_of_assortative_mating * variance_if_identical + (1-strength_of_assortative_mating)*variance_if_independent
}

# Check the math. These should all be roughly the same.
computeVariance(0.5, 0.5)
var(generateAlleleCounts(0.5, 0.5, 100000))
computeVariance(0.3, 0.3)
var(generateAlleleCounts(0.3, 0.3, 100000))
computeVariance(0.3, 0)
var(generateAlleleCounts(0.3, 0, 100000))
computeVariance(0.3, 1)
var(generateAlleleCounts(0.3, 1, 100000))

# How I generated the data for the last question
library("magrittr")
filenames = paste0("demo", 1:9, ".csv")
for(f in filenames){
  am = sample(c(0, 0.3), 1)
  sapply( (1:9)/10, function(af) generateAlleleCounts(af, am, 1000) ) %>% 
    as.data.frame %>% 
    set_colnames(paste0("Locus", 1:9)) %>% 
    write.csv(f)
}

File: course content/4_mistaken_assumptions/assortative_mating.py
import numpy as np
import pandas as pd

def generate_allele_counts(allele_frequency, strength_of_assortative_mating, n):
    # Check user input -- should be numbers from 0 to 1
    assert isinstance(allele_frequency, (int, float)) and 0 <= allele_frequency <= 1, f"Allele frequency must be between 0 and 1; got {allele_frequency}"
    assert isinstance(strength_of_assortative_mating, (int, float)) and 0 <= strength_of_assortative_mating <= 1, f"Strength of assortative mating must be between 0 and 1; got {strength_of_assortative_mating}"

    extreme_assortative_mating = 2 * np.random.binomial(1, allele_frequency, n)
    no_assortative_mating = np.random.binomial(2, allele_frequency, n)

    return np.where(np.random.binomial(1, strength_of_assortative_mating, n),
                    extreme_assortative_mating,
                    no_assortative_mating)

def compute_variance(allele_frequency, strength_of_assortative_mating):
    # Check user input -- should be numbers from 0 to 1
    assert isinstance(allele_frequency, (int, float)) and 0 <= allele_frequency <= 1, f"Allele frequency must be between 0 and 1; got {allele_frequency}"
    assert isinstance(strength_of_assortative_mating, (int, float)) and 0 <= strength_of_assortative_mating <= 1, f"Strength of assortative mating must be between 0 and 1; got {strength_of_assortative_mating}"

    variance_if_independent = 2 * allele_frequency * (1 - allele_frequency)
    variance_if_identical = 4 * allele_frequency * (1 - allele_frequency)

    return strength_of_assortative_mating * variance_if_identical + (1 - strength_of_assortative_mating) * variance_if_independent

# Check the math. These should all be roughly the same.
print(compute_variance(0.5, 0.5))
print(np.var(generate_allele_counts(0.5, 0.5, 100000)))
print(compute_variance(0.3, 0.3))
print(np.var(generate_allele_counts(0.3, 0.3, 100000)))
print(compute_variance(0.3, 0))
print(np.var(generate_allele_counts(0.3, 0, 100000)))
print(compute_variance(0.3, 1))
print(np.var(generate_allele_counts(0.3, 1, 100000)))

# How I generated the data for the last question
filenames = [f"demo{i}.csv" for i in range(1, 10)]
for i, filename in enumerate(filenames):
    am = np.random.choice([0, 0.3], 1)[0]
    allele_counts = np.array([generate_allele_counts(af, am, 1000) for af in np.arange(0.1, 1, 0.1)]).T
    allele_counts_df = pd.DataFrame(allele_counts, columns=[f"Locus{i}" for i in range(1, 10)])
    allele_counts_df.to_csv(filename, index=False)

# How to read a CSV into Python as a dataframe
# If you have a local file called "demo.csv"
allele_counts = pd.read_csv("demo1.csv")
# If you want to pull it straight from the course website using a URL
allele_counts = pd.read_csv("https://raw.githubusercontent.com/ekernf01/HEART_choosing_stat_methods/main/course%20content/4_mistaken_assumptions/demo_data/demo1.csv")
